dbms/src/Functions/FunctionsRegexp.h

```diff
@@ -0,0 +1,944 @@
+};
+
+// Implementation of regexp and regexp_like functions
+template <typename Name>
+class FunctionStringRegexp : public FunctionStringRegexpBase
+    , public IFunction
+{
+public:
+    using ResultType = UInt8;
+    static constexpr auto name = Name::name;
+
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionStringRegexp>(); }
+    String getName() const override { return name; }
+    bool isVariadic() const override { return true; }
+    void setCollator(const TiDB::TiDBCollatorPtr & collator_) override { collator = collator_; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        size_t args_max_num;
+        constexpr std::string_view class_name_sv(Name::name);
+        constexpr std::string_view regexp_like_name_sv(NameRegexpLike::name);
+
+        if constexpr (class_name_sv == regexp_like_name_sv)
+            args_max_num = REGEXP_LIKE_MAX_PARAM_NUM;
+        else
+            args_max_num = REGEXP_MAX_PARAM_NUM;
+
+        size_t arg_num = arguments.size();
+        if (arg_num < REGEXP_XXX_MIN_PARAM_NUM || arg_num > args_max_num)
+            throw Exception("Illegal argument number");
+
+        bool has_nullable_col = false;
+        bool has_data_type_nothing = false;
+
+        for (size_t i = 0; i < REGEXP_XXX_MIN_PARAM_NUM; ++i)
+            checkInputArg(arguments[i], &has_nullable_col, &has_data_type_nothing);
+
+        // check match_type arg for regexp_like
+        if constexpr (class_name_sv == regexp_like_name_sv)
+            if (arg_num == args_max_num && !arguments[args_max_num - 1]->isString())
+                checkInputArg(arguments[args_max_num - 1], &has_nullable_col, &has_data_type_nothing);
+
+        if (has_nullable_col)
+        {
+            if (has_data_type_nothing)
+                return std::make_shared<DataTypeNullable>(std::make_shared<DataTypeNothing>());
+            return std::make_shared<DataTypeNullable>(std::make_shared<DataTypeNumber<ResultType>>());
+        }
+        else
+        {
+            if (has_data_type_nothing)
+                return std::make_shared<DataTypeNothing>();
+            return std::make_shared<DataTypeNumber<ResultType>>();
+        }
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        // Do something related with nullable columns
+        NullPresence null_presence = getNullPresense(block, arguments);
+
+        const ColumnPtr & col_expr = block.getByPosition(arguments[0]).column;
+
+        if (null_presence.has_const_null_col || null_presence.has_data_type_nothing)
+        {
+            // There is a const null column in the input
+            block.getByPosition(result).column = block.getByPosition(result).type->createColumnConst(block.rows(), Null());
```

Comment:
I'm not sure if this line can work if the return type if `DataTypeNullable(DataTypeNothing)`